climada_petals/hazard/copernicus_forecast/handler.py

```diff
@@ -0,0 +1,560 @@
+1. CDS API client installation:
+   pip install cdsapi
+
+2. CDS account and API key:
+   Register at https://cds-beta.climate.copernicus.eu
+
+3. CDS API configuration:
+   Create a .cdsapirc file in your home directory with your API key and URL.
+   For instructions, visit:
+   https://cds-beta.climate.copernicus.eu/how-to-api#install-the-cds-api-client
+"""
+
+import os
+import logging
+import calendar
+import re
+
+import xarray as xr
+import pandas as pd
+import numpy as np
+import cdsapi
+
+from climada.hazard import Hazard
+from climada.util.coordinates import get_country_geometries
+import climada_petals.hazard.copernicus_forecast.indicator as indicator
+
+
+LOGGER = logging.getLogger(__name__)
+
+class ForecastHandler:
+    """
+    A class to handle downloading, processing, and calculating climate indices
+    and hazards based on seasonal forecast data from Copernicus Climate Data Store (CDS).
+    """
+
+    _FORMAT_GRIB = 'grib'
+    _FORMAT_NC = 'nc'
+    
+    def __init__(self, data_dir='.', url = None, key = None):
+        """
+        Initializes the ForecastHandler instance.
+
+        Parameters:
+        data_dir (str): Path to the directory where downloaded and processed data will be stored.
+            Defaults to the current directory ('.').
+        url (str): url to the CDS API. Defaults to None, in which case the url from /.cdsapirc
+            is used.
+        key (str): CDS API key to the CDS API. Defaults to None, in which case the key from
+            /.cdsapirc is used.
+
+        Note:
+        This method sets up logging and initializes the data directory for the instance.
+        """
+        logging.basicConfig(format='%(asctime)s | %(levelname)s : %(message)s', level=logging.INFO)
+        self.logger = logging.getLogger()
+        self.data_dir = data_dir
+        self.key = key
+        self.url = url
+    
+    @staticmethod
+    def _get_bounds_for_area_selection(area_selection, margin=0.2):
+        """
+        Determines the geographic bounds based on an area selection string.
+
+        Parameters:
+        area_selection (str): Specifies the area for data selection.
+        margin (float): Additional margin to be added to the bounds in degrees.
+
+        Returns:
+        list: A list of four floats representing the bounds [north, east, south, west].
+        """
+        if isinstance(area_selection, str):
+            if area_selection.lower() == "global":
+                return [90, -180, -90, 180]  # north, west, south, east
+        else:
+            # try if area was given in bounds
+            try: 
+                north, west, south, east = area_selection
+                lat_margin = margin * (north - south)
+                lon_margin = margin * (east - west)
+                north += lat_margin
+                east += lon_margin
+                south -= lat_margin
+                west -= lon_margin
+                return [north, west, south, east]
+            except:
```

Comment:
NEVER raise an exception and not handle it.

Reply:
Done